typing/includemod.ml

```diff
@@ -389,9 +388,39 @@ and try_modtypes ~in_eq ~loc env ~mark subst mty1 mty2 =
   | (Mty_alias p1, Mty_alias p2) ->
       if Env.is_functor_arg p2 env then
         Error (Error.Invalid_module_alias p2)
-      else if not (equal_module_paths env p1 subst p2) then
+      else
+        let p2' = Subst.module_path subst p2 in
+        if not (equal_module_paths env p1 p2') then
           Error Error.(Mt_core Incompatible_aliases)
-      else Ok Tcoerce_none
+        else if Path.same p2 p2' then
+          Ok Tcoerce_none
+        else
+          (* This path contains local references, we may need to build a
+             coercion.
```

Comment:
Interestingly, I think we could avoid this problem if we had transparent ascription. The underlying issue is that given:
```ocaml
struct module A : S module B = A end
: sig module A' : S' module B' = A' end
```
(where `'` is just used to distinguish the two sides) we produce a substitution like:
```ocaml
A' :-> A
```
but the correct substitution is really:
```ocaml
A' :-> (A <: S')
```
which would automatically give the right coercion in these cases.

Once we have transparent ascription that may be a better approach, although it requires some care around recursive modules (`S'` might refer to `A'`, but not in aliases, so those needs to be substituted out first using the current substitution) and also some care to ensure efficiency.